### PR TITLE
gh-103951: enable optimization for fast attribute access on module subclasses

### PR DESCRIPTION
**Pull request by skirpichev**
This patch relax specialization requirements from
``PyModule_CheckExact(op)`` to
``Py_TYPE(op)->tp_getattro != PyModule_Type.tp_getattro``.

Benchmarks:

```py
# a.py
import pyperf
import b
import c
runner = pyperf.Runner()
runner.timeit('b.x', 'b.x', globals=globals())
runner.timeit('c.x', 'c.x', globals=globals())
```
```py
# b.py
x = 1
```
```py
# c.py
import sys, types
x = 1
class _Foo(types.ModuleType): pass
sys.modules[__name__].__class__ = _Foo
```

On the main:
```
$ python a.py -q
b.x: Mean +- std dev: 50.2 ns +- 2.7 ns
c.x: Mean +- std dev: 132 ns +- 7 ns
```

With the patch:
```
$ python a.py -q
b.x: Mean +- std dev: 52.9 ns +- 3.6 ns
c.x: Mean +- std dev: 52.6 ns +- 2.7 ns
```

* Issue: gh-103951

### CODE REVIEW

**Comment by skirpichev** [comment]
JIT failure seems unrelated, see similar https://github.com/python/cpython/actions/runs/11599941049 in the main.

File: Include/internal/pycore_moduleobject.h [hunk 1]
```diff
@@ -16,6 +16,12 @@ extern int _PyModule_IsPossiblyShadowing(PyObject *);
 
 extern int _PyModule_IsExtension(PyObject *obj);
 
+static inline int
+_PyModule_HasSpecializedGetAttr(PyObject* op)
+{
+    return Py_TYPE(op)->tp_getattro != PyModule_Type.tp_getattro;
```
Comment (Fidget-Spinner) [inline Include/internal/pycore_moduleobject.h]: I'm thinking whether it's better to just leave this as is and not wrap it in a function. WDYT?

Comment (skirpichev) [reply Include/internal/pycore_moduleobject.h]: Do you think (1) it's better to make it a macro?  Or (2) just inline that code?

Comment (Fidget-Spinner) [reply Include/internal/pycore_moduleobject.h]: I think (2)

**Review by Fidget-Spinner — APPROVED** [review]

**Comment by Fidget-Spinner** [comment]
@skirpichev do you have commit rights?

**Comment by skirpichev** [comment]
@Fidget-Spinner, no.  I'm just member of Python triage team.

**Pull request merged**